uber/models.py

```diff
@@ -2086,7 +2089,7 @@ def register_session_listeners():
     """
     listen(Session.session_factory, 'before_flush', _acquire_badge_lock)
     listen(Session.session_factory, 'before_flush', _presave_adjustments)
-    listen(Session.session_factory, 'before_flush', _track_changes)
+    listen(Session.session_factory, 'after_flush', _track_changes)
```

Comment:
one slightly non-obvious but extremely important thing to pay attention to here:  we're messing locking c.BADGE_LOCK in the before_flush call and unlocking it in the after_flush call.  You're going to want to make sure that the stuff in _track_changes still happens after _acquire_badge_lock() is called and before _release_badge_lock() is called.

If lock ordering is messed with here, here be dragons.

I _think_ what you're doing here is sane, but to be safe I would step through this in the debugger to make super-sure.   Problems with the order of these functions were the root cause of our occasional deadlocks we were seeing previously (which all seem to be fixed at the moment)